src/copy_stubs.c

```diff
@@ -0,0 +1,203 @@
+
+/* ----------------------------------------------- */
+/* Clone a file given input and output fd          */
+/* It must fully complete or fully fail.           */
+
+#if defined(__linux__) || defined(__linux)
+
+
+#include <sys/ioctl.h>
+
+#if !defined(FICLONE) && defined(_IOW)
+#define FICLONE _IOW(0x94, 9, int)
+#endif
+
+CAMLprim value unsn_clone_file(value in_fd, value out_fd)
+{
+  CAMLparam2(in_fd, out_fd);
+
+#ifdef FICLONE
+  caml_release_runtime_system();
+  int status = ioctl(Int_val(out_fd), FICLONE, Int_val(in_fd));
+  caml_acquire_runtime_system();
+
+  /* Don't raise an exception, just return false in case of errors */
+  CAMLreturn(Val_bool(status == 0));
+#else /* defined(FICLONE) */
+  CAMLreturn(Val_false);
+#endif
+}
+
+
+#else /* defined(__linux__) */
+
+
+CAMLprim value unsn_clone_file(value in_fd, value out_fd)
+{
+  CAMLparam2(in_fd, out_fd);
+  CAMLreturn(Val_false);
+}
+
+
+#endif /* defined(__linux__) */
+
+
+/* --------------------------------------------------------- */
+/* Copy, or possibly clone, a file given input and output fd */
+/* If operation is not supported by the OS or the filesystem
+   then file offsets must not have been changed at failure.
+   Output file offset must be changed on success.            */
+
+#if defined(__linux__) || defined(__linux)
+
+
+#include <unistd.h>
+#include <sys/syscall.h>
+#include <sys/sendfile.h>
+
+CAMLprim value unsn_copy_file(value in_fd, value out_fd, value offs, value len)
+{
+  CAMLparam4(in_fd, out_fd, offs, len);
+  off_t off_i = Int64_val(offs);
+  ssize_t ret;
+
+  caml_release_runtime_system();
+#ifdef __NR_copy_file_range
+  /* First, try copy_file_range() */
+  /* Using off_i prevents changing in_fd file offset */
+  ret = syscall(__NR_copy_file_range, Int_val(in_fd), &off_i, Int_val(out_fd), NULL, Long_val(len), 0);
+  if (ret == -1 && (errno == ENOSYS || errno == EBADF || errno == EXDEV))
+#endif /* defined(__NR_copy_file_range) */
+  {
+    /* Second, try sendfile(); this one changes out_fd file offset */
+    ret = sendfile(Int_val(out_fd), Int_val(in_fd), &off_i, Long_val(len));
+  }
+  caml_acquire_runtime_system();
+  if (ret == -1) caml_uerror("copy_file", Nothing);
+
+  CAMLreturn(Val_long(ret));
+}
+
+
+#elif defined(__sun) || defined(sun)
+
+
+#include <sys/sendfile.h>
+
+CAMLprim value unsn_copy_file(value in_fd, value out_fd, value offs, value len)
+{
+  CAMLparam4(in_fd, out_fd, offs, len);
+  off_t off = orig_off = Int64_val(offs);
+  ssize_t ret;
+
+  caml_release_runtime_system();
+  /* This one changes out_fd file offset */
+  ret = sendfile(Int_val(out_fd), Int_val(in_fd), &off, Long_val(len));
+  caml_acquire_runtime_system();
+  if (ret == -1) {
+    if (off > off_orig) {
+      ret = off - off_orig;
+    } else {
+      caml_uerror("copy_file", Nothing);
+    }
+  }
+
+  CAMLreturn(Val_long(ret));
+}
+
+
+#else /* defined(__linux__) || defined(__sun) */
+
+
+CAMLprim value unsn_copy_file(value in_fd, value out_fd, value offs, value len)
```

Comment:
This looks like it fails always because of lack of support for better-than-copy.  Whether I guessed right or not a comment would be good.

Reply:
Yes, it's something of an idiom. I added comments on expected return values and exceptions, so I included this too.